common/src/main/java/io/pravega/common/io/ByteBufferOutputStream.java

```diff
@@ -0,0 +1,183 @@
+/**
+ * Copyright (c) Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.common.io;
+
+import io.pravega.common.io.serialization.RandomAccessOutputStream;
+import io.pravega.common.util.BufferView;
+import io.pravega.common.util.ByteArraySegment;
+import java.io.OutputStream;
+import java.nio.ByteBuffer;
+import javax.annotation.concurrent.NotThreadSafe;
+
+/**
+ * An {@link OutputStream} that is backed by a {@link ByteBuffer} which can be resized as needed. Prefer using this
+ * instead of {@link java.io.ByteArrayOutputStream} since this makes direct use of {@link ByteBuffer} intrinsic methods
+ * for writing primitive types (see {@link DirectDataOutput} for example).
+ */
+@NotThreadSafe
+public class ByteBufferOutputStream extends OutputStream implements RandomAccessOutputStream {
+    //region Private
+
+    private static final int MAX_ARRAY_SIZE = Integer.MAX_VALUE - Long.BYTES;
+    private ByteBuffer buf;
+
+    //endregion
+
+    //region Constructor
+
+    /**
+     * Creates a new instance of the {@link ByteBufferOutputStream}.
+     */
+    public ByteBufferOutputStream() {
+        this(128);
+    }
+
+    /**
+     * Creates a new instance of the {@link ByteBufferOutputStream} with given initial size.
+     *
+     * @param initialSize The initial size to create with.
+     */
+    public ByteBufferOutputStream(int initialSize) {
+        this.buf = ByteBuffer.allocate(initialSize);
+    }
+
+    //endregion
+
+    //region Implementation
+
+    @Override
+    public void write(int b) {
+        ensureExtraCapacity(1);
+        this.buf.put((byte) b);
+    }
+
+    @Override
+    public void write(int byteValue, int streamPosition) {
```

Comment:
The contract for `RandomAccessOutputStream` (where this method is defined) explicitly states that `streamPosition` must be within bounds, otherwise an `IndexOutOfBoundsException` will be thrown back. 

The consuming code (of that interface) takes advantage of this, for example, in serialization, when it needs to "go back" and back-fill the length field after the serialization is over. If this had not been possible/allowed, its only choice would have been to precompute that length before the serialization, a potentially unnecessary step. By allowing this back-fill, all that the serializer needs to do is count the number of bytes written (as they are written) and then fill that position upon close.

Reply:
I understood what the code was doing. I was pointing out there is a better way to handle this usecase, and I already wrote the code for it. Take a look at that pattern.